packages/components/src/Table/Row/TableRow.js

```diff
@@ -0,0 +1,174 @@
+	if (rowDataGetter && rowDataGetter.getElementId) {
+		return rowDataGetter.getElementId(element);
+	} else if (element.id && typeof element.id === 'string') {
+		return element.id;
+	}
+	return index.toString();
+}
+
+function getRowDataClassName(classNameProvider, element, columnKey) {
+	if (classNameProvider && classNameProvider.getForRowData) {
+		return classNameProvider.getForRowData(columnKey, element);
+	}
+	return `tc-table-row-data-${columnKey}`;
+}
+
+function getCellComponent(rowRenderer, columnKey) {
+	if (rowRenderer && rowRenderer.getCellComponent) {
+		return rowRenderer.getCellComponent(columnKey);
+	}
+	return TableCell;
+}
+
+function getCellComponentExtraProps(rowRenderer, columnKey) {
+	if (rowRenderer && rowRenderer.getExtraProps) {
+		return rowRenderer.getExtraProps(columnKey);
+	}
+	return null;
+}
+
+function getRowData(rowDataGetter, element, columnKey) {
+	if (rowDataGetter && rowDataGetter.getRowData) {
+		return rowDataGetter.getRowData(element, columnKey);
+	} else if (element && element[columnKey]) {
+		return element[columnKey];
+	}
+	return null;
+}
+
+/**
+ * This function is responsible for rendering a piece of data for an element.
+ */
+function renderRowData(element, index, columnKey, rowDataGetter, classNameProvider, rowRenderer) {
+	const CellComponent = getCellComponent(rowRenderer, columnKey);
+	const compKey = `${getRowId(rowDataGetter, element, index)}-${columnKey}`;
+	const classnames = classNames(`td-${columnKey}`, theme['tc-table-row-cell']);
+	const dataClassnames = classNames(
+		'tc-table-row-data',
+		theme['tc-table-row-data'],
+		getRowDataClassName(classNameProvider, element, columnKey),
+	);
+	return (
+		<td key={`td-${compKey}`} className={classnames}>
+			<CellComponent
+				key={compKey}
+				element={element}
+				data={getRowData(rowDataGetter, element, columnKey)}
+				className={dataClassnames}
+				extra={getCellComponentExtraProps(rowRenderer, columnKey)}
+			/>
+		</td>
+	);
+}
+
+function getRowClassName(classNameProvider, element) {
+	if (classNameProvider && classNameProvider.getForRow) {
+		return classNameProvider.getForRow(element);
+	}
+	return '';
+}
+
+/**
+ * This component displays the data of an element in a table.
+ * A row is divided in columns, each column displaying an element data.
+ */
+export default class TableRow extends Component {
+	constructor(props) {
+		super(props);
+		this.handleMouseEnter = this.handleMouseEnter.bind(this);
+		this.handleMouseLeave = this.handleMouseLeave.bind(this);
+		this.updateRowRef = this.updateRowRef.bind(this);
+	}
+
+	componentDidMount() {
+		if (this.rowRef != null) {
+			this.rowRef.addEventListener('mouseenter', this.handleMouseEnter);
+			this.rowRef.addEventListener('mouseleave', this.handleMouseLeave);
+		}
+	}
+
+	shouldComponentUpdate(nextProps) {
+		if (this.props.rowRenderer && this.props.rowRenderer.needRowUpdate) {
+			return this.props.rowRenderer.needRowUpdate(nextProps);
```

Comment:
What is the need behind that ?
Why do we let an external function control the rendering of the component ?

Reply:
We could control the rendering of the rows depending on the parent state for example. This feature will be used by the data-mapper.

Reply:
I still don't get it. I understand that something from the parent controls if this component should render or not, but I don't understand why. This pattern is a little weird.

Reply:
well in fact i need it for some performance issues in the future data-mapper. When i display hundreds of rows, i provide some custom optimizations in order to reduce the rendering process.